### PR TITLE
[ModelFactory] Implemented Chat factory

### PR DESCRIPTION
**Pull request by kinelski**
Adding the Model Factory for Chat models. This factory can be used to instantiate models for mocking.

At the moment there's no way to have the factories generated automatically since they are split across multiple namespaces.

I tried to keep them as close to generated code as possible. In case we manage to have the code generator generating them automatically for us, it would be ideal if we could keep changes to a minimum when migrating from manually written to generated code.

The following models have not been included to the factory since they are also used as input for the client and can already be instantiated by costumers:
- ChatMessage
- ChatMessageContentPart
- ChatFunctionCall
- ChatToolCall

### CODE REVIEW

File: .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    {
+        InternalChatCompletionMessageToolCallChunkFunction function = new InternalChatCompletionMessageToolCallChunkFunction(
+            functionName,
+            functionArgumentsUpdate,
+            serializedAdditionalRawData: null);
+
+        return new StreamingChatToolCallUpdate(
+            index,
+            id,
+            kind,
```
Comment (kinelski) [inline .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs:171]: I'm allowing customers to set `kind` here but at the moment we only support the value `Function`. Another possibility would be removing this parameter altogether and forcibly setting `Kind` to `Function`.

Comment (joseharriaga) [reply .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs:171]: This is fine!

File: .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+using System.Linq;
+
+namespace OpenAI.Chat;
+
+/// <summary> Model factory for models. </summary>
+public static partial class OpenAIChatModelFactory
+{
+    /// <summary> Initializes a new instance of <see cref="OpenAI.Chat.ChatCompletion"/>. </summary>
+    /// <returns> A new <see cref="OpenAI.Chat.ChatCompletion"/> instance for mocking. </returns>
+    public static ChatCompletion ChatCompletion(
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs:12]: FYI: I believe ChatCompletion has new properties from the Structured Outputs feature, so you might need to do a few updates.

**Review by joseharriaga — APPROVED** [review]

File: .dotnet/api/OpenAI.netstandard2.0.cs [hunk 3]
```diff
@@ -1473,6 +1473,15 @@ public class FunctionChatMessage : ChatMessage, IJsonModel<FunctionChatMessage>,
         public static bool operator !=(ImageChatMessageContentPartDetail left, ImageChatMessageContentPartDetail right);
         public override readonly string ToString();
     }
+    public static class OpenAIChatModelFactory {
```
Comment (joseharriaga) [inline .dotnet/api/OpenAI.netstandard2.0.cs:1476]: Before merging the PR, please update the CHANGELOG too. 😊

File: .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs (hunk 2 above)
Comment (kinelski) [reply .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs:12]: Thanks for letting me know. I have updated the code with the new properties (`Refusal` and `RefusalTokenLogProbabilities`). Are these the ones you're talking about? I couldn't find others.

Comment (joseharriaga) [reply .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs:12]: Yes, thank you!

**Review by joseharriaga — APPROVED** [review]

**Pull request merged**